sc-poc/modules/pubsub/websocket.go

```diff
@@ -0,0 +1,165 @@
+	channels := a.Channels()
+	apis := []*apis.API{}
+
+	for path, channel := range channels.Channels {
+		// Get the publish and subscribe API of the channel
+		publisherAPI := a.getPublisherAPI(path, channel.Name)
+		subscriberAPI := a.getSubscriberAPI(path, channel.Name)
+
+		apis = append(apis, publisherAPI, subscriberAPI)
+	}
+
+	return apis
+}
+
+// getPublishAPI creates a websocket API for sending messages in the channel
+func (a *App) getPublisherAPI(channelPath, channelName string) *apis.API {
+	return &apis.API{
+		Name: fmt.Sprintf("%s-publisher", channelName),
+		Path: fmt.Sprintf("/v1/pubsub/default%s/publisher", channelPath),
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// Create the websocket connection
+			conn, _, _, err := ws.UpgradeHTTP(r, w)
+			if err != nil {
+				a.logger.Error("could not establish websocket connection", zap.String("channel", channelName), zap.Error(err))
+				return
+			}
+
+			// Go routine to receive messages from the websocket connection and publish
+			// them into the watermill channel
+			go func() {
+				defer conn.Close()
+				for {
+					// Get the message from the websocket connection
+					data, _, err := wsutil.ReadClientData(conn)
+					if err != nil {
+						a.logger.Error("could not read client data or the connection is closed", zap.String("channel", channelName), zap.Error(err))
+						return
+					}
+
+					var message Message
+					err = json.Unmarshal(data, &message)
+					if err != nil {
+						a.logger.Error("could not unmarshal data", zap.String("channel", channelName), zap.Error(err))
+						continue
+					}
+
+					if message.Event == messageEvent {
+						var pubMsg PublishMessage
+						err = mapstructure.Decode(message.Data, &pubMsg)
+						if err != nil {
+							a.logger.Error("could not decode data", zap.String("channel", channelName), zap.Error(err))
+							continue
+						}
+
+						if pubMsg.ID == "" {
+							pubMsg.ID = uuid.NewString()
+						}
+
+						err = a.Publish(channelName, pubMsg, PublishOptions{})
+						if err != nil {
+							a.logger.Error("could not publish client message", zap.String("channel", channelName), zap.Error(err))
+						}
+					}
+				}
+			}()
+		}),
+	}
+}
+
+// getSubscriberAPI creates a websocket API for receiving messages from the channel
+func (a *App) getSubscriberAPI(channelPath, channelName string) *apis.API {
+	return &apis.API{
+		Name: fmt.Sprintf("%s-subscriber", channelName),
+		Path: fmt.Sprintf("/v1/pubsub/default%s/subscriber", channelPath),
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ctx, cancel := context.WithCancel(r.Context())
+			defer cancel()
+
+			// Create the websocket connection
+			conn, _, _, err := ws.UpgradeHTTP(r, w)
+			if err != nil {
+				a.logger.Error("could not establish websocket connection", zap.String("channel", channelName), zap.Error(err))
+				return
+			}
+
+			// Wait for subscribe event
+			for {
```

Comment:
This implementation won't work cause it doesn't leave room for you to check for the unsubscribe event that could come in next

Reply:
Should we use a variable which stores the state(subscribed/not subscribed) in which the client is in right now?